torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -54,20 +54,72 @@ static IValue Table(
 }
 
 std::string getModulePath(Node* node) {
-  std::string modulePath = node->scopeName();
-  size_t end = modulePath.size();
-  // Here we remove the source range information to make the
-  // module debugging information shorter and cleaner.
-  if (modulePath[end - 1] == '>') {
-    end = modulePath.rfind('<');
-    if (end > 0 && modulePath[end - 1] == '<') {
-      --end;
+  auto callStack = node->callstack();
+  return callStack.has_value() ? callStack->get()->getModulePath() : "";
+}
+
+std::string getModuleTypeName(const Module& module, const std::string& prefix) {
+  std::string moduleType = module.type()->str();
+  size_t lastDotIndex = moduleType.rfind('.');
+  if (lastDotIndex != std::string::npos) {
+    moduleType = moduleType.substr(lastDotIndex + 1);
+  }
+  return prefix + "(" + moduleType + ")";
+}
+
+void setModulePath(Node* node, const std::string& root_scope_string) {
+  if (!node->callstack()) {
+    std::string root = root_scope_string + ".forward";
+    InlinedCallStackPtr inlineCallStack =
+        c10::make_intrusive<InlinedCallStack>();
+    node->setCallStack(inlineCallStack);
+    node->setModulePathStr(root);
+  } else {
+    std::string module_info = root_scope_string;
+    auto callstack_ptr = *(node->callstack());
+    const auto& vec = callstack_ptr->vec();
+
+    for (size_t i = 0; i < vec.size(); ++i) {
+      const auto& tup = vec[i];
+      const auto opt_module_instance_info = std::get<2>(tup);
+      if (opt_module_instance_info) {
+        const auto& module_instance_info = opt_module_instance_info.value();
+        if (module_instance_info.class_type()) {
+          const auto& class_type = module_instance_info.class_type();
+          const auto& instance_name = module_instance_info.instance_name();
+          auto type_name = class_type->name()->qualifiedName();
+          type_name = type_name.substr(type_name.find_last_of(".") + 1);
+          module_info += "." + instance_name + "(" + type_name + ")";
+          if (i == vec.size() - 1) {
+            module_info += "." + std::get<0>(tup)->name();
+          }
+        } else {
+          module_info += ".(UNKNOWN_INSTANCE(UNKNOWN_TYPE)";
+        }
+      } else {
+        module_info += ".(UNKNOWN_INSTANCE(UNKNOWN_TYPE)";
+      }
+    }
+
+    node->setModulePathStr(module_info);
```

Comment:
So what I am thinking is that instead of doing this we can just have `std::unordered_map<Node*, std::string> node_module_paths;` and basically we just do `node_module_paths.emplace(node, module_info)`. Then you `getModulePath` will be `return node_module_paths[node]` in which case you may not need that function at all. You can have node_module_path created in getFunctionTuple and passed down from setModulePaths.

Reply:
That's a good idea. I tried it and it seems not working out of box. After looking into it, the nodes generated from `code.instructions_source()[i]` are from `PreprocessGraph`, which is a graph copy and the node pointer will be different. If it's expected, what will be the ideal key to find the unique node?

Reply:
I see. Yes that is bit troubling. One thing I can think of is moving these lines, https://github.com/pytorch/pytorch/blob/5dad1d541229f97d386967eb177baa5c1740e374/torch/csrc/jit/serialization/export_module.cpp#L132-L135, here: https://github.com/pytorch/pytorch/blob/5dad1d541229f97d386967eb177baa5c1740e374/torch/csrc/jit/serialization/export_module.cpp#L147. So instead of doing 
```
  if (save_mobile_debug_info) {
    std::string root_scope_string = getModuleTypeName(module, "top");
    setModulePaths(graph, root_scope_string);
  }
```
you actually do
```
  if (save_mobile_debug_info) {
    std::string root_scope_string = getModuleTypeName(module, "top");
    setModulePath(node, root_scope_string);
  }
```
I think some variant of this should work, since we are not really using graph to populate debug info. Instead we are using `Node`'s callstack to populate debug info. So if you can just use node to build module_debug_info that is even better. We do not have to maintain any map, we get the debug info right when we need it.